File: 0173_Binary Search Tree Iterator/BinarySearchTreeIterator.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class BSTIterator {

    private Stack<TreeNode> stack = new Stack<TreeNode>();
    
    public BSTIterator(TreeNode root) {
        PushAllLeftChild(root);
    }
    
    /** @return the next smallest number */
    public int Next() {
        var current = stack.Pop();
        PushAllLeftChild(current.right);
        return current.val;
    }
    
    /** @return whether we have a next smallest number */
    public bool HasNext() {
        return stack.Count > 0;
    }
    
    private void PushAllLeftChild(TreeNode node){
        var current = node;
        while(current!=null){
            stack.Push(current);
            current = current.left;
        }
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.Next();
 * bool param_2 = obj.HasNext();
 */

File: 0173_Binary Search Tree Iterator/BinarySearchTreeIterator.java
class BSTIterator {
    private TreeNode _root;
    private Stack<TreeNode> _stack;

    public BSTIterator(TreeNode root) {
        _root = root;
        _stack = new Stack<TreeNode>();
    }
    
    public int next() {
        int v = 0;
        
        while(_root!=null) {
            _stack.push(_root);
            _root = _root.left;
        }
        
        if(_stack.size() > 0) {
            _root = _stack.pop();
            v = _root.val;
            _root = _root.right;
        } 
        return v;
    }
    
    public boolean hasNext() {
        return _root!=null || _stack.size() > 0;
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */